### ISSUE TITLE
iri unable to resolve DNS

### ISSUE THREAD

**Issue by MarcoGlauser** [iotaledger/iri#479]
I'm running iri in a kubernetes cluster with multiple replicas and add all the replicas as neighbors of eachother. Since the cluster internal ip is not guaranteed to be stable, kubernetes offers a dns resolver.
When i try to start iri with the dns entries provided by kubernetes as neighbor I get the following Error:
``` 
01/09 16:01:36.773 [main] ERROR com.iota.iri.network.Node - 'udp://iota-node-0.iota-node-internal.iota.svc.cluster.local:14600' is not a valid uri schema or resolvable address.
01/09 16:01:36.795 [main] ERROR com.iota.iri.network.Node - 'udp://iota-node-1.iota-node-internal.iota.svc.cluster.local:14600' is not a valid uri schema or resolvable address.
01/09 16:01:36.819 [main] ERROR com.iota.iri.network.Node - 'udp://iota-node-2.iota-node-internal.iota.svc.cluster.local:14600' is not a valid uri schema or resolvable address.
```
The DNS entry is perfectly fine and i can access the node manually through that entry. When I add the neighbors by IP-Address, iri connects to them without any problem.
Can someone point me to the direction where I can fix this and make a PR?

**Comment by ahab94** [comment]
Add `--dns-resolution-false` flag to ignore DNS refresher thread.

**Comment by MarcoGlauser** [comment]
I still get the same error message and the neigbors won't be added
Here the complete log with the --dns-resolution-false flag.
```
01/10 06:46:48.595 [main] INFO  com.iota.iri.IRI - Remote access enabled. Binding API socket to listen any interface.
01/10 06:46:48.598 [main] INFO  com.iota.iri.IRI - Welcome to IRI 1.4.1.6
01/10 06:46:55.562 [main] INFO  c.i.i.s.r.RocksDBPersistenceProvider - Initializing Database Backend... 
01/10 06:47:06.556 [main] INFO  c.i.i.s.r.RocksDBPersistenceProvider - RocksDB persistence provider initialized.
01/10 06:47:06.571 [main] INFO  com.iota.iri.network.UDPReceiver - UDP replicator is accepting connections on udp port 14600
01/10 06:47:06.572 [UDP receiving thread] INFO  com.iota.iri.network.UDPReceiver - Spawning Receiver Thread
01/10 06:47:06.572 [UDP receiving thread] INFO  com.iota.iri.network.UDPReceiver - Receiver thread processed/dropped ratio: 0/0
01/10 06:47:06.579 [main] INFO  c.i.i.network.replicator.Replicator - Started ReplicatorSourcePool
01/10 06:47:06.584 [Thread-2] INFO  c.i.i.n.r.ReplicatorSourcePool - TCP replicator is accepting connections on tcp port 15600
01/10 06:47:06.601 [main] ERROR com.iota.iri.network.Node - 'udp://iota-node-0.iota-node-internal.iota.svc.cluster.local:14600' is not a valid uri schema or resolvable address.
01/10 06:47:06.621 [main] ERROR com.iota.iri.network.Node - 'udp://iota-node-1.iota-node-internal.iota.svc.cluster.local:14600' is not a valid uri schema or resolvable address.
01/10 06:47:06.629 [main] ERROR com.iota.iri.network.Node - 'udp://iota-node-2.iota-node-internal.iota.svc.cluster.local:14600' is not a valid uri schema or resolvable address.
01/10 06:47:06.647 [pool-2-thread-1] INFO  com.iota.iri.network.Node - Spawning Broadcaster Thread
01/10 06:47:06.657 [pool-2-thread-2] INFO  com.iota.iri.network.Node - Spawning Tips Requester Thread
01/10 06:47:06.667 [pool-2-thread-3] INFO  com.iota.iri.network.Node - Ignoring DNS Refresher Thread... DNS_RESOLUTION_ENABLED is false
01/10 06:47:06.676 [pool-2-thread-4] INFO  com.iota.iri.network.Node - Spawning Process Received Data Thread
01/10 06:47:06.691 [pool-2-thread-5] INFO  com.iota.iri.network.Node - Spawning Reply To Request Thread
01/10 06:47:06.751 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 0 , toReply = 0 / totalTransactions = 7874659
01/10 06:47:06.980 [main] INFO  com.iota.iri.IRI - IOTA Node initialised correctly. 
```
The log says `Ignoring DNS Refresher Thread... DNS_RESOLUTION_ENABLED is false`.

**Comment by ahab94** [comment]
Probably the function that verifies `uri` addresses, needs some adjustments. Looking into it.

**Comment by ahab94** [comment]
@MarcoGlauser, so the following is the code  
```JAVA
 public boolean isUriValid(final URI uri) {
        if (uri != null) {
            if (uri.getScheme().equals("tcp") || uri.getScheme().equals("udp")) {
                if ((new InetSocketAddress(uri.getHost(), uri.getPort()).getAddress() != null)) {
                    return true;
                }
            }
            log.error("'{}' is not a valid uri schema or resolvable address.", uri);
            return false;
        }
        log.error("Cannot read uri schema, please check neighbor config!");
        return false;
    }
```
That end up giving you

```BASH
01/09 16:01:36.773 [main] ERROR com.iota.iri.network.Node - 'udp://iota-node-0.iota-node-internal.iota.svc.cluster.local:14600' is not a valid uri schema or resolvable address.
```
The problem is that `InetSocketAddress(...).getAddress()` is returning null because it is supposed to return null if the ip or hostname is unresolved. Thus your scenario.
```
 /**
     *
     * Gets the {@code InetAddress}.
     *
     * @return the InetAdress or {@code null} if it is unresolved.
     */
    public final InetAddress getAddress() {
        return holder.getAddress();
    }
```
What i suspect is that your pods is not being able to resolve each other by their `hostname`.
Could you share your kubernetes manifest?

**Comment by MarcoGlauser** [comment]
@ahab94 Thank you very much for your help! 
I was able to resolve the issue. It was a problem within kubernetes. The problem was that the pod itself could not resolve the dns entry because it uses the hostnetwork. Somehow kubernetes is not able to use the default `dnsPolicy:ClusterFirst` in that case.
Fortunately the option `ClusterFirstWithHostNet` was added in kubernetes 1.6 which solves this issue.
I'll release the complete kubernetes manifest in a seperate repo once i get it stable.
